upgrader.go

```diff
@@ -64,6 +64,14 @@ func (u *Upgrader) UpgradeInbound(ctx context.Context, t transport.Transport, ma
 	return u.upgrade(ctx, t, maconn, "", network.DirInbound)
 }
 
+// UpgradeInboundWithPeerCheck upgrades the given inbound multiaddr-net connection into a
+// full libp2p-transport connection.
+// It verifies that the remote peer ID matches p.
+// This is needed when upgrading a connection that resulted from a TCP simultaneous connect.
+func (u *Upgrader) UpgradeInboundWithPeerCheck(ctx context.Context, t transport.Transport, maconn manet.Conn, p peer.ID) (transport.CapableConn, error) {
```

Comment:
This doesn't do anything. Why do we need it?

Reply:
Oh, I see. Is this really necessary? We can just as easily check _after_ the fact.

Reply:
We can do it either way. I don’t have a strong opinion on this, I just thought it would be nice to abort as early as possible.

Reply:
Alternatively, how about exposing `upgrade` as `Upgrade` and deprecating the others?

NOTE: the current patch won't work because `setupSecurity` assumes that "no peer id" means "inbound".

Reply:
You're right. If we want to kill the connection on the TLS / noise level, this needs to be wired through the `SecureTransport` / `SecureMuxer` interfaces.